runtime/src/iree/builtins/ukernel/arch/arm_64/mmt4d_arm_64_dotprod.c

```diff
@@ -75,3 +75,135 @@ IREE_UK_MMT4D_TILE_FUNC_IMPL_FOR_M0_1_2_4_8(
+      lhs[2] = vget_high_s8(lhs_uzp_0.val[0]);
+      lhs[3] = vget_high_s8(lhs_uzp_0.val[1]);
+      lhs[4] = vget_low_s8(lhs_uzp_1.val[0]);
+      lhs[5] = vget_high_s8(lhs_uzp_1.val[0]);
+      lhs[6] = vget_low_s8(lhs_uzp_1.val[1]);
+      lhs[7] = vget_high_s8(lhs_uzp_1.val[1]);
+    } else if (M0 == 4) {
+      int8x16x2_t lhs_uzp = vld2q_s8(lhs_ptr);
+      lhs_ptr += 32;
+      lhs[0] = vget_low_s8(lhs_uzp.val[0]);
+      lhs[1] = vget_low_s8(lhs_uzp.val[1]);
+      lhs[2] = vget_high_s8(lhs_uzp.val[0]);
+      lhs[3] = vget_high_s8(lhs_uzp.val[1]);
+    } else if (M0 == 2) {
+      int8x8x2_t lhs_uzp = vld2_s8(lhs_ptr);
+      lhs_ptr += 16;
+      lhs[0] = lhs_uzp.val[0];
+      lhs[1] = lhs_uzp.val[1];
+    } else {  // M0 == 1.
+      int8x8_t r = vld1_s8(lhs_ptr);
+      lhs_ptr += 8;
+      int8x8x2_t lhs_uzp = vuzp_s8(r, vzero);
+      lhs[0] = lhs_uzp.val[0];
+      lhs[1] = lhs_uzp.val[1];
+    }
+
+    // 1x2 * 2x8 -> 1x8.
+    acc[0] = vdotq_lane_s32(acc[0], rhs[0], lhs[0], 0);
+    acc[1] = vdotq_lane_s32(acc[1], rhs[1], lhs[0], 0);
+    acc[0] = vdotq_lane_s32(acc[0], rhs[2], lhs[1], 0);
+    acc[1] = vdotq_lane_s32(acc[1], rhs[3], lhs[1], 0);
+    if (M0 == 1) continue;
+    // 2x2 * 2x8 -> 2x8.
+    acc[2] = vdotq_lane_s32(acc[2], rhs[0], lhs[0], 1);
+    acc[3] = vdotq_lane_s32(acc[3], rhs[1], lhs[0], 1);
+    acc[2] = vdotq_lane_s32(acc[2], rhs[2], lhs[1], 1);
+    acc[3] = vdotq_lane_s32(acc[3], rhs[3], lhs[1], 1);
+    if (M0 == 2) continue;
+    // 4x2 * 2x8 -> 4x8.
+    acc[4] = vdotq_lane_s32(acc[4], rhs[0], lhs[2], 0);
+    acc[5] = vdotq_lane_s32(acc[5], rhs[1], lhs[2], 0);
+    acc[6] = vdotq_lane_s32(acc[6], rhs[0], lhs[2], 1);
+    acc[7] = vdotq_lane_s32(acc[7], rhs[1], lhs[2], 1);
+    acc[4] = vdotq_lane_s32(acc[4], rhs[2], lhs[3], 0);
+    acc[5] = vdotq_lane_s32(acc[5], rhs[3], lhs[3], 0);
+    acc[6] = vdotq_lane_s32(acc[6], rhs[2], lhs[3], 1);
+    acc[7] = vdotq_lane_s32(acc[7], rhs[3], lhs[3], 1);
+    if (M0 == 4) continue;
+    // 8x2 * 2x8 -> 8x8.
+    acc[8] = vdotq_lane_s32(acc[8], rhs[0], lhs[4], 0);
+    acc[9] = vdotq_lane_s32(acc[9], rhs[1], lhs[4], 0);
+    acc[10] = vdotq_lane_s32(acc[10], rhs[0], lhs[4], 1);
+    acc[11] = vdotq_lane_s32(acc[11], rhs[1], lhs[4], 1);
+    acc[12] = vdotq_lane_s32(acc[12], rhs[0], lhs[5], 0);
+    acc[13] = vdotq_lane_s32(acc[13], rhs[1], lhs[5], 0);
+    acc[14] = vdotq_lane_s32(acc[14], rhs[0], lhs[5], 1);
+    acc[15] = vdotq_lane_s32(acc[15], rhs[1], lhs[5], 1);
+    acc[8] = vdotq_lane_s32(acc[8], rhs[2], lhs[6], 0);
+    acc[9] = vdotq_lane_s32(acc[9], rhs[3], lhs[6], 0);
+    acc[10] = vdotq_lane_s32(acc[10], rhs[2], lhs[6], 1);
+    acc[11] = vdotq_lane_s32(acc[11], rhs[3], lhs[6], 1);
+    acc[12] = vdotq_lane_s32(acc[12], rhs[2], lhs[7], 0);
+    acc[13] = vdotq_lane_s32(acc[13], rhs[3], lhs[7], 0);
+    acc[14] = vdotq_lane_s32(acc[14], rhs[2], lhs[7], 1);
+    acc[15] = vdotq_lane_s32(acc[15], rhs[3], lhs[7], 1);
```

Comment:
Please un-unroll this. Something like

```c
for (a) for (b) for (c) {
  acc[8 + 4 * b + a] = vdotq_lane_s32(acc[8 + 4 * b + a], rhs[a], lhs[4 + b], 0);
  acc[8 + 4 * b + a + 2] = vdotq_lane_s32(acc[8 + 4 * b + a], rhs[a + 2 * c], lhs[4 + b], 1);
}
```

Reply:
(and likewise the other cases above).

Reply:
I don't think that pseudocode works. If we take this example has the inner loop:

```
acc[8] = vdotq_lane_s32(acc[8], rhs[0], lhs[4], 0);
acc[10] = vdotq_lane_s32(acc[10], rhs[0], lhs[4], 1);
```

Both rhs indexes are the same but the pseudocode has `a` for the first line, and `a + 2 * c` for the second line.

I re-arranged the code so that it is calculating acc's that are 2 indexes apart with alternating lanes:

```
acc[8] = vdotq_lane_s32(acc[8], rhs[0], lhs[4], 0);
acc[10] = vdotq_lane_s32(acc[10], rhs[0], lhs[4], 1);

acc[9] = vdotq_lane_s32(acc[9], rhs[1], lhs[4], 0);
acc[11] = vdotq_lane_s32(acc[11], rhs[1], lhs[4], 1);

acc[12] = vdotq_lane_s32(acc[12], rhs[0], lhs[5], 0);
acc[14] = vdotq_lane_s32(acc[14], rhs[0], lhs[5], 1);

acc[13] = vdotq_lane_s32(acc[13], rhs[1], lhs[5], 0);
acc[15] = vdotq_lane_s32(acc[15], rhs[1], lhs[5], 1);

acc[8] = vdotq_lane_s32(acc[8], rhs[2], lhs[6], 0);
acc[10] = vdotq_lane_s32(acc[10], rhs[2], lhs[6], 1);

acc[9] = vdotq_lane_s32(acc[9], rhs[3], lhs[6], 0);
acc[11] = vdotq_lane_s32(acc[11], rhs[3], lhs[6], 1);

acc[12] = vdotq_lane_s32(acc[12], rhs[2], lhs[7], 0);
acc[14] = vdotq_lane_s32(acc[14], rhs[2], lhs[7], 1);

acc[13] = vdotq_lane_s32(acc[13], rhs[3], lhs[7], 0);
acc[15] = vdotq_lane_s32(acc[15], rhs[3], lhs[7], 1);
```

And I cannot get this down to 3 `for` loops with 2 lines in the body. Best I can do is 1 for loop with 8 lines in the body:

```
for (int i = 0; i < 2; i++) {
  acc[8] = vdotq_lane_s32(acc[8], rhs[2 * i + 0], lhs[4 + 2 * i + 0], 0);
  acc[9] = vdotq_lane_s32(acc[9], rhs[2 * i + 0 + 1], lhs[4 + 2 * i + 0], 0);
  acc[10] = vdotq_lane_s32(acc[10], rhs[2 * i + 0], lhs[4 + 2 * i + 0], 1);
  acc[11] = vdotq_lane_s32(acc[11], rhs[2 * i + 0 + 1], lhs[4 + 2 * i + 0], 1);
  acc[12] = vdotq_lane_s32(acc[12], rhs[2 * i + 0], lhs[4 + 2 * i + 1], 0);
  acc[13] = vdotq_lane_s32(acc[13], rhs[2 * i + 0 + 1], lhs[4 + 2 * i + 1], 0);
  acc[14] = vdotq_lane_s32(acc[14], rhs[2 * i + 0], lhs[4 + 2 * i + 1], 1);
  acc[15] = vdotq_lane_s32(acc[15], rhs[2 * i + 0 + 1], lhs[4 + 2 * i + 1], 1);
}
```

Let me know if you can find something that works.

Reply:
Brilliant! That works. Refactored the other parts too.